### PR TITLE
[Bugfix][Sampler] Keep exactly k tokens in top-k with tied logits

### PR DESCRIPTION
**Pull request by liuxiang-sudo**
## Summary

The PyTorch top-k paths (`apply_top_k_top_p_pytorch` and `apply_top_k_only`) used a **strict threshold comparison** (`logits < k-th largest`), which keeps **all** tokens tied at the k-th largest logit — potentially more than `k` tokens. The Triton kernel keeps **exactly `k`** entries (truncating the tied group).

Because `apply_top_k_top_p` dispatches by batch size (`logits.shape[0] >= 8` → Triton, else PyTorch), the same request can produce different sampled tokens depending on whether it is run alone or in a batch of >= 8. This contradicts the project's determinism goals (`tests/v1/determinism/`).

Ties in model logits are not rare: with bf16 weights (e.g. Qwen3-0.6B), ties within the top-8 occur at a large fraction of sampled positions.

This aligns the PyTorch paths with the Triton kernel (`torch.topk` semantics: keep exactly `k`, truncate ties).

## Why this is not duplicating an existing PR

Searched GitHub for `top_k tie`, `top_k duplicate`, and related terms — no existing issue or PR covers this top-k tie-handling divergence between the PyTorch and Triton paths. The existing test `tests/v1/sample/test_topk_topp_sampler.py` asserts the paths match for random logits, where exact ties essentially never occur, so this divergence is not covered.

## Changes

- `vllm/v1/sample/ops/topk_topp_sampler.py`:
  - `apply_top_k_top_p_pytorch`: mask positions below the top-k (keep exactly `k` of the sorted logits) instead of threshold comparison.
  - `apply_top_k_only`: keep exactly the first `k` of `topk.indices` instead of threshold comparison.
- `tests/v1/sample/test_topk_topp_sampler.py`: add regression tests that construct tied logits and assert both paths keep exactly `k` entries.

## Test plan

```bash
# Before the fix, these new tests fail (Triton keeps 2, PyTorch kept 4+):
.venv/bin/python -m pytest tests/v1/sample/test_topk_topp_sampler.py -q -k "ties or exactly_k"
# -> 4 failed

# After the fix:
.venv/bin/python -m pytest tests/v1/sample/test_topk_topp_sampler.py -q
# -> 133 passed, 3 skipped
```

Full file: `128 passed, 3 skipped` before adding the new tests; `133 passed, 3 skipped` after.

Lint:
```bash
pre-commit run ruff-check --files vllm/v1/sample/ops/topk_topp_sampler.py tests/v1/sample/test_topk_topp_sampler.py  # Passed
pre-commit run ruff-format --files vllm/v1/sample/ops/topk_topp_sampler.py tests/v1/sample/test_topk_topp_sampler.py  # Passed
```

## Model evaluation

Sampling behavior for non-tied logits is unchanged (both paths already kept the same `k` entries when no ties were present at the boundary). No model weights or inference kernels were modified.

## AI assistance disclosure

This PR was developed with the assistance of an AI coding agent (opencode). All changed lines were reviewed by a human; tests were run locally as documented above.

### CODE REVIEW

**Review by NolenLiang — CHANGES_REQUESTED** [review]
This makes all implementations retain exactly `k` entries, but it does not yet make tied-logit sampling batch invariant because the implementations can still retain different *token IDs*.

The Triton kernel resolves a tie according to its vocabulary scan/reduction order. The two PyTorch paths now rely on `logits.sort(..., stable=False)` and `torch.topk(...).indices`; PyTorch does not guarantee stable indices for tied elements, and those two paths are not guaranteed to agree with each other or with Triton. Since dispatch changes with batch size, equal finite counts are insufficient: the sampling support, processed logprobs, and sampled token can still change across the batch-size boundary.

The new tests miss this because they only compare `(result != -inf).sum(...)`. The top-k+top-p test does not even independently assert the final count is `k`.

Please define one deterministic tie-break (for example, lower token ID wins in triton) and apply it consistently in the PyTorch sort path, the PyTorch topk-only path, and Triton. Tests should compare the actual finite mask/token IDs through the public dispatch for batch sizes on both sides of the path boundary (e.g. 1/7/8), not just counts. Ideally reuse fixed exponential noise and assert the sampled token is also identical; include the sampling-mask/replay path because this difference affects RL distribution replay.

Minor: `k.to(torch.long).clamp_(...)` can mutate a caller-owned tensor when `k` is already int64 and needlessly bumps its version counter. Please use non-in-place `.clamp(...)` here.

**Comment by new-TonyWang** [comment]
I tested it with tied logits and confirmed that the  this PR keeps exactly k tokens and resolves the inconsistency between the PyTorch and Triton paths. This pr looks great to me ! @aoshen02
